missing permission causes app to crash

if user denies camera permission and tries capturing an image or record video the app crashes
```
FATAL EXCEPTION: main
Process: org.ole.planet.myplanet, PID: 16318
java.lang.SecurityException: Permission Denial: starting Intent { act=android.media.action.IMAGE_CAPTURE flg=0x3 cmp=com.android.camera2/com.android.camera.CaptureActivity clip={text/uri-list hasLabel(0) {U(content)}} (has extras) } from ProcessRecord{6b2598e 16318:org.ole.planet.myplanet/u0a239} (pid=16318, uid=10239) with revoked permission android.permission.CAMERA
	at android.os.Parcel.createExceptionOrNull(Parcel.java:3057)
	at android.os.Parcel.createException(Parcel.java:3041)
	at android.os.Parcel.readException(Parcel.java:3024)
	at android.os.Parcel.readException(Parcel.java:2966)
	at android.app.IActivityTaskManager$Stub$Proxy.startActivity(IActivityTaskManager.java:2110)
	at android.app.Instrumentation.execStartActivity(Instrumentation.java:1873)
	at android.app.Activity.startActivityForResult(Activity.java:5589)
	at androidx.activity.ComponentActivity.startActivityForResult(ComponentActivity.java:780)
	at androidx.core.app.ActivityCompat.startActivityForResult(ActivityCompat.java:243)
	at androidx.activity.ComponentActivity$1.onLaunch(ComponentActivity.java:239)
	at androidx.activity.result.ActivityResultRegistry$2.launch(ActivityResultRegistry.java:173)
	at androidx.fragment.app.Fragment$10.launch(Fragment.java:3612)
	at androidx.activity.result.ActivityResultLauncher.launch(ActivityResultLauncher.java:47)
	at org.ole.planet.myplanet.ui.resources.AddResourceFragment.takePhoto(AddResourceFragment.kt:243)
	at org.ole.planet.myplanet.ui.resources.AddResourceFragment.onCreateView$lambda$10(AddResourceFragment.kt:132)
	at org.ole.planet.myplanet.ui.resources.AddResourceFragment.$r8$lambda$j8IkNRDTlUQAlt7H5KQg-81mFVI(Unknown Source:0)
	at org.ole.planet.myplanet.ui.resources.AddResourceFragment$$ExternalSyntheticLambda11.onClick(D8$$SyntheticClass:0)
	at android.view.View.performClick(View.java:7659)
	at android.view.View.performClickInternal(View.java:7636)
	at android.view.View.-$$Nest$mperformClickInternal(Unknown Source:0)
	at android.view.View$PerformClick.run(View.java:30156)
	at android.os.Handler.handleCallback(Handler.java:958)
	at android.os.Handler.dispatchMessage(Handler.java:99)
	at android.os.Looper.loopOnce(Looper.java:205)
	at android.os.Looper.loop(Looper.java:294)
	at android.app.ActivityThread.main(ActivityThread.java:8177)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:971)
Caused by: android.os.RemoteException: Remote stack trace:
	at com.android.server.wm.ActivityTaskSupervisor.checkStartAnyActivityPermission(ActivityTaskSupervisor.java:1136)
	at com.android.server.wm.ActivityStarter.executeRequest(ActivityStarter.java:1069)
	at com.android.server.wm.ActivityStarter.execute(ActivityStarter.java:742)
	at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1288)
	at com.android.server.wm.ActivityTaskManagerService.startActivityAsUser(ActivityTaskManagerService.java:1239)

```

--- a/app/build.gradle
+++ b/app/build.gradle
@@ -10,8 +10,8 @@ android {
         applicationId "org.ole.planet.myplanet"
         minSdkVersion 26
         targetSdkVersion 34
-        versionCode 2490
-        versionName "0.24.90"
+        versionCode 2501
+        versionName "0.25.1"
         ndkVersion '21.3.6528147'
         testInstrumentationRunner "androidx.test.runner.AndroidJUnitRunner"
         vectorDrawables.useSupportLibrary = true

--- a/app/src/main/java/org/ole/planet/myplanet/ui/resources/AddResourceFragment.kt
+++ b/app/src/main/java/org/ole/planet/myplanet/ui/resources/AddResourceFragment.kt
@@ -60,6 +60,7 @@ class AddResourceFragment : BottomSheetDialogFragment() {
     private lateinit var captureVideoLauncher: ActivityResultLauncher<Uri>
     private lateinit var openFolderLauncher: ActivityResultLauncher<String>
     private lateinit var requestPermissionLauncher: ActivityResultLauncher<String>
+    private lateinit var requestCameraLauncher: ActivityResultLauncher<String>
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         if (arguments != null) {
@@ -106,6 +107,30 @@ class AddResourceFragment : BottomSheetDialogFragment() {
                         .show()
                 } else {
                     Utilities.toast(requireContext(), "Microphone permission is required to record audio.")
+                }
+            }
+        }
+        requestCameraLauncher = registerForActivityResult(
+            ActivityResultContracts.RequestPermission()
+        ) { isGranted ->
+            if (isGranted) {
+                takePhoto()
+            } else {
+                if (!shouldShowRequestPermissionRationale(Manifest.permission.RECORD_AUDIO)) {
+                    AlertDialog.Builder(requireContext(), R.style.AlertDialogTheme)
+                        .setTitle(R.string.permission_required)
+                        .setMessage(R.string.camera_permission_required)
+                        .setPositiveButton(R.string.settings) { dialog, _ ->
+                            dialog.dismiss()
+                            val intent = Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS)
+                            val uri: Uri = Uri.fromParts("package", requireContext().packageName, null)
+                            intent.data = uri
+                            startActivity(intent)
+                        }
+                        .setNegativeButton(R.string.cancel) { dialog, _ -> dialog.dismiss() }
+                        .show()
+                } else {
+                    Utilities.toast(requireContext(), "camera permission is required.")
                 }
             }
         }
@@ -214,6 +239,11 @@ class AddResourceFragment : BottomSheetDialogFragment() {
     }
 
     private fun dispatchTakeVideoIntent() {
+        if (ContextCompat.checkSelfPermission(requireContext(), Manifest.permission.CAMERA)
+            != PackageManager.PERMISSION_GRANTED){
+            requestCameraLauncher.launch(Manifest.permission.CAMERA)
+            return
+        }
         val takeVideoIntent = Intent(MediaStore.ACTION_VIDEO_CAPTURE)
         videoUri = createVideoFileUri()
         takeVideoIntent.putExtra(MediaStore.EXTRA_OUTPUT, videoUri)
@@ -232,6 +262,11 @@ class AddResourceFragment : BottomSheetDialogFragment() {
     }
 
     private fun takePhoto() {
+        if (ContextCompat.checkSelfPermission(requireContext(), Manifest.permission.CAMERA)
+            != PackageManager.PERMISSION_GRANTED){
+            requestCameraLauncher.launch(Manifest.permission.CAMERA)
+            return
+        }
         val values = ContentValues().apply {
             put(MediaStore.Images.Media.TITLE, "Photo_" + UUID.randomUUID().toString())
             put(MediaStore.Images.Media.MIME_TYPE, "image/jpeg")

--- a/app/src/main/res/values-ar/strings.xml
+++ b/app/src/main/res/values-ar/strings.xml
@@ -1296,5 +1296,6 @@
     <string name="permission_required">مطلوب إذن</string>
     <string name="microphone_permission_required">يلزم الحصول على إذن الميكروفون لتسجيل الصوت. يُرجى تفعيله من إعدادات التطبيق</string>
     <string name="confirm_share_community">هل أنت متأكد أنك تريد مشاركة هذا المحتوى مع المجتمع؟</string>
+    <string name="camera_permission_required">يلزم الحصول على إذن الكاميرا. يُرجى تفعيله من إعدادات التطبيق.</string>
 
 </resources>

--- a/app/src/main/res/values-es/strings.xml
+++ b/app/src/main/res/values-es/strings.xml
@@ -1296,5 +1296,6 @@
     <string name="permission_required">Se requiere permiso</string>
     <string name="microphone_permission_required">Se requiere permiso del micrófono para grabar audio. Actívalo en la configuración de la aplicación.</string>
     <string name="confirm_share_community">¿Estás seguro de que quieres compartir este contenido con la comunidad?</string>
+    <string name="camera_permission_required">Se requiere permiso para usar la cámara. Actívalo en la configuración de la aplicación.</string>
 
 </resources>

--- a/app/src/main/res/values-fr/strings.xml
+++ b/app/src/main/res/values-fr/strings.xml
@@ -1296,5 +1296,6 @@
     <string name="permission_required">Autorisation requise</string>
     <string name="microphone_permission_required">L\'autorisation du microphone est requise pour l\'enregistrement audio. Veuillez l\'activer dans les paramètres de l\'application.</string>
     <string name="confirm_share_community">Êtes-vous sûr de vouloir partager ce contenu avec la communauté ?</string>
+    <string name="camera_permission_required">L\'autorisation de l\'appareil photo est requise. Veuillez l\'activer dans les paramètres de l\'application.</string>
 
 </resources>

--- a/app/src/main/res/values-ne/strings.xml
+++ b/app/src/main/res/values-ne/strings.xml
@@ -1296,5 +1296,6 @@
     <string name="permission_required">अनुमति आवश्यक छ</string>
     <string name="microphone_permission_required">अडियो रेकर्ड गर्न माइक्रोफोन अनुमति आवश्यक छ। कृपया एप सेटिङहरूमा गई यसलाई सक्षम पार्नुहोस्।</string>
     <string name="confirm_share_community">के तपाईं यो सामग्री समुदायसँग साझा गर्न चाहनुहुन्छ?</string>
+    <string name="camera_permission_required">क्यामेरा अनुमति आवश्यक छ। कृपया एप सेटिङहरूमा गई यसलाई सक्षम गर्नुहोस्।</string>
 
 </resources>

--- a/app/src/main/res/values-so/strings.xml
+++ b/app/src/main/res/values-so/strings.xml
@@ -1296,5 +1296,6 @@
     <string name="permission_required">Ogolaanshaha ayaa loo baahan yahay</string>
     <string name="microphone_permission_required">Ogolaanshaha makarafoonka ayaa loo baahan yahay si loo duubo codka Fadlan awood u geli goobaha abka</string>
     <string name="confirm_share_community">Ma hubtaa inaad rabto inaad la wadaagto macluumaadkan bulshada?</string>
+    <string name="camera_permission_required">ogolaanshaha kamarada ayaa loo baahan yahay. Fadlan awood u geli goobaha abka</string>
 
 </resources>

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -1296,5 +1296,5 @@
     <string name="permission_required">Permission Required</string>
     <string name="microphone_permission_required">Microphone permission is required to record audio. Please enable it in the app settings.</string>
     <string name="confirm_share_community">Are you sure you want to share this content with the community?</string>
-
+    <string name="camera_permission_required">Camera permission is required. Please enable it in the app settings.</string>
 </resources>

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -1297,4 +1297,5 @@
     <string name="microphone_permission_required">Microphone permission is required to record audio. Please enable it in the app settings.</string>
     <string name="confirm_share_community">Are you sure you want to share this content with the community?</string>
     <string name="camera_permission_required">Camera permission is required. Please enable it in the app settings.</string>
+    
 </resources>

--- a/app/src/main/res/values/strings.xml
+++ b/app/src/main/res/values/strings.xml
@@ -1297,5 +1297,5 @@
     <string name="microphone_permission_required">Microphone permission is required to record audio. Please enable it in the app settings.</string>
     <string name="confirm_share_community">Are you sure you want to share this content with the community?</string>
     <string name="camera_permission_required">Camera permission is required. Please enable it in the app settings.</string>
-    
+
 </resources>

